src/openssl.c

```diff
@@ -11975,6 +11975,47 @@ static int cipher_final(lua_State *L) {
 } /* cipher_final() */
 
 
+static int cipher_get_tag(lua_State *L) {
+	EVP_CIPHER_CTX *ctx = checksimple(L, 1, CIPHER_CLASS);
+	luaL_Buffer tag;
+	size_t tag_size = luaL_checkinteger(L, 2);
+
+	luaL_buffinit(L, &tag);
+
+	if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, tag_size, (void*)luaL_prepbuffsize(&tag, tag_size))) {
```

Comment:
When was `EVP_CIPHER_CTX_ctrl` introduced? luaossl still tries to be compatible with old openssl versions

Reply:
I believe it's been around a "long time". I don't know if it was in 0.9.8 (I can dig if really necessary), but it's certainly in 1.0.2. It's due to be retired post-3.0.

Reply:
> It's due to be retired post-3.0.

What is it going to be replaced with?

Reply:
A new `OSSL_PARAM` API they've [introduced](https://beta.openssl.org/docs/manmaster/man3/OSSL_PARAM.html). I haven't looked too much into it so far, as I haven't been working on any code that can be 3.0.x-only just yet.